Use `ChainedLocalRepositoryManager` to support `-Dmaven.repo.local.tail`

tiagobento opened the pull request:
Hi Quarkus team! 👋 My first PR here! ~(still draft as I didn't have time to go through all the contribution guidelines yet)~

---

`maven-resolver` introduced [Chained LRM](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=235837219) a while back. This PR adds support to it, making it possible for Quarkus Maven Plugin to find dependencies on local repositories declared as tail.

Tested locally with a project I have and it worked great, but still need to research how to create tests for it and update the relevant parts of the documentation.

All feedback is appreciated. Thanks!

tiagobento committed:
Use ChainedLocalRepositoryManager

geoand commented:
Thanks a lot for the contribution!

@aloubyansky can you have a look?

aloubyansky commented on independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/BootstrapMavenContext.java:
```diff
@@ -460,9 +479,15 @@ private DefaultRepositorySystemSession newRepositorySystemSession() throws Boots
+            tail.add(getRepositorySystem().newLocalRepositoryManager(session, new LocalRepository(tailPath)));
+        }
+
+        session.setLocalRepositoryManager(new ChainedLocalRepositoryManager(head, tail, getLocalRepoTailIgnoreAvailability()));
```
Let's do this only if the tail was actually configured. Also for now let's not introduce public methods that are used only in this same class.
Thanks!

tiagobento replied:
Thanks! Done in ed070ce.

tiagobento commented:
@aloubyansky Any pointers on how to test this? And on what parts of the documentation need to be updated?

aloubyansky commented:
I guess create a couple of local Maven repos in some temp dir and init the resolver, such as in https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/resolver/maven/test/BootstrapMavenContextTestBase.java?

tiagobento reviewed (commented):
@aloubyansky Added tests and fixed a few bugs that the tests helped me catch :)

I took the liberty to make a small change to `BootstrapMavenContextTestBase.java` to make it easier to configure `BootstrapMavenContext`, and had to adapt another test (`PreferPomsFromWorkspaceTest.java`)

I gave some thought to the documentation changes, and maybe there isn't a need to update anything, given `-Dmaven.repo.local.tail` is a feature of Maven itself, so actually it would be expected that it works out of the box with Quarkus. From this perspective, this PR is more like a "bug fix" then adding new stuff.

Let me know what you think. I converted the PR to "ready for review" :)

Thanks!

aloubyansky reviewed (commented):
Thanks @tiagobento 
Would you mind adding a test for multiple locations in the tail? Perhaps you want to use different artifacts for that. Also if the same version of the artifact is found in multiple locations, I suppose the first location in the list will win?
BTW, you could use `txt` artifacts instead of actual binaries for test artifacts.

tiagobento commented:
Added two more tests:
- `testValidTailResolutionOrder`; and
- `testValidTailMultiplicity`

---

Fixed the ArrayList creation with a fixed length.

---

Kept the `.jar` extension since they're only empty files. I can change to `.txt`, though. Just let me know.

aloubyansky commented on independent-projects/bootstrap/maven-resolver/src/test/java/io/quarkus/bootstrap/resolver/maven/test/ChainedLocalRepositoryManagerTest.java:
```diff
@@ -0,0 +1,206 @@
+                BootstrapMavenContext.config()
+                        .setLocalRepositoryTail(new String[] { M2_FROM_REMOTE }));
+
+        assertNotNull(resolveOrgAcmeFooJar001(mvn));
```
@tiagobento could you please explain this one. `ignoreAvailability` is `true`. Is it availability of artifacts that should be ignored?

tiagobento replied:
Sure. So maybe the `ignoreAvailability` name kind of gives the wrong idea of what it actually does, and maybe a more explicit, albeit verbose, name would be `ignoreRemoteAvailability`? cc @cstamas

This flag is only applicable for dependencies that have been downloaded from a remote repository prior to being consumed as part of a `tail` local repository. See the `_remote.repositories` file.

This particular test succeeds because `maven.repo.local.tail.ignoreAvailability` ends up being `true`, so even though org.acme:foo:jar:0.0.1 was supposedly initially resolved from a remote repository with id = `some-repo-id`, this fact is ignored, and it being available in the local repository suffices. When `maven.repo.local.tail.ignoreAvailability` is false, it fails. That case is exactly what is covered by `testValidTailFromRemoteCheckingForAvailabilityViaSystemProp`.

aloubyansky replied:
Thanks for the explanation. The property name is indeed very confusing. It sounds like the meaning is "trust any remote repo" that provided an artifact.

cstamas replied:
"availability" in resolver means that tracked remote repository (by repoID) is available in the context artifact is asked for. So "remote" may come in as "remote repo ID is available in current context asking for artifact" :)

aloubyansky commented:
@tiagobento could you please squash the commits? Thanks.

tiagobento commented:
@aloubyansky Done. Thanks!

tiagobento commented:
@aloubyansky Thanks for merging it! 

Is there a chance to get this back-ported to 3.15? I see the tag is already created for 3.15.0, but maybe for 3.15.1? My team is waiting for 3.15 to be out so we can upgrade from 3.8. Having this feature on the LTS release would be very valuable to us 🙏 Let me know and I'll send the PR if that's ok. Thanks again!

aloubyansky commented:
I think it can be backported. I added a label, it'll be considered.
FYI, I opened https://github.com/quarkusio/quarkus/pull/43420 as a follow up.

Thank you and congratulations on your first contribution to Quarkus! It's a nice one, I'll use it in our tests.

tiagobento commented:
@aloubyansky Thanks!

tiagobento commented:
@aloubyansky Hi! I saw this missed 3.15.1 :( Any change it can get included in 3.15.2?

aloubyansky commented:
It has the backport label, so yes, it will be considered

tiagobento commented:
@gsmet Understood. I'll keep an eye on 3.16 and the upcoming 3.15 patches. Thanks a lot for considering it and getting back to me here ☺️
